lib/src/native/realm_core.dart

```diff
@@ -500,52 +502,45 @@ class _RealmCore {
+      return;
+    }
+
+    syncConfig.syncErrorHandler(syncError);
+  }
+
+  static void _callResetCallback(FutureOr<void> Function() callback, Pointer<Void> unlockCallbackFunc) async {
+    bool success = true;
+    try {
+      await callback();
+    } catch (error) {
+      success = false;
+      _realmLib.realm_register_user_code_callback_error(error.toPersistentHandle());
+    } finally {
+      _realmLib.realm_dart_invoke_unlock_callback(success, unlockCallbackFunc);
     }
   }
 
   static void _syncBeforeResetCallback(Object userdata, Pointer<shared_realm> realmHandle, Pointer<Void> unlockCallbackFunc) {
     final syncConfig = userdata as FlexibleSyncConfiguration;
-    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset;
-    if (beforeResetCallback == null) {
-      _invokeNativeFunction(unlockCallbackFunc);
-      return;
-    }
+    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset!;
+
     final realm = RealmInternal.getUnowned(syncConfig, RealmHandle._unowned(realmHandle));
-    _continueWhenComplete(() => beforeResetCallback(realm), unlockCallbackFunc);
+
+    _callResetCallback(() => beforeResetCallback(realm), unlockCallbackFunc);
   }
 
   static void _syncAfterResetCallback(Object userdata, Pointer<shared_realm> beforeHandle, Pointer<realm_thread_safe_reference> afterReference, bool didRecover,
       Pointer<Void> unlockCallbackFunc) {
     final syncConfig = userdata as FlexibleSyncConfiguration;
-    final afterResetCallback = didRecover ? syncConfig.clientResetHandler.onAfterRecovery : syncConfig.clientResetHandler.onAfterDiscard;
-    if (afterResetCallback == null) {
-      _invokeNativeFunction(unlockCallbackFunc);
-      return;
-    }
+    final afterResetCallback = didRecover ? syncConfig.clientResetHandler.onAfterRecovery! : syncConfig.clientResetHandler.onAfterDiscard!;
+
     final beforeRealm = RealmInternal.getUnowned(syncConfig, RealmHandle._unowned(beforeHandle));
     final afterRealm = RealmInternal.getUnowned(
       syncConfig,
       RealmHandle._unowned(_realmLib.realm_from_thread_safe_reference(afterReference, scheduler.handle._pointer)),
     );
-    _continueWhenComplete(() => afterResetCallback(beforeRealm, afterRealm), unlockCallbackFunc);
-  }
-
-  static void _invokeNativeFunction(Pointer<Void> nativeFunction, [Object? error]) {
-    if (error != null) {
-      _realmLib.realm_register_user_code_callback_error(error.toPersistentHandle());
-    }
-    _realmLib.realm_dart_invoke_native_with_result(error == null, nativeFunction);
-  }
 
-  static Future<void> _continueWhenComplete(FutureOr<void> Function() callback, Pointer<Void> nativeFunction) async {
-    try {
-      await callback();
-      _invokeNativeFunction(nativeFunction);
-    } catch (error) {
-      _invokeNativeFunction(nativeFunction, error);
-    }
+     _callResetCallback(() => afterResetCallback(beforeRealm, afterRealm), unlockCallbackFunc);
```

Comment:
Could we find some better name? There are too much uses of the name `ResetCallbacks`. It is not clear what we do in this method.

Reply:
`_callResetCallback` calls the reset call backs in a specific way and should not be used in any other place in the SDK. Do you have any suggestion for a better name?